src/middleware/request-id/request-id.ts

```diff
@@ -0,0 +1,62 @@
+/**
+ * @module
+ * Request ID Middleware for Hono.
+ */
+
+import type { MiddlewareHandler } from '../../types'
+
+export type RequestIDVariables = {
+  requestID: string
+}
+
+export type RequesIDOptions = {
+  limitLength?: number
+  headerName?: string
+  generator?: () => string
+}
+
+/**
+ * Request ID Middleware for Hono.
+ *
+ * @param {object} options - Options for Request ID middleware.
+ * @param {number} [options.limitLength=255] - The maximum length of request id.
+ * If positive truncates the request id at the specified length.
+ * @param {string} [options.headerName=X-Request-Id] - The header name used in request id.
+ * @param {generator} [options.generator=crypto.randomUUID()] - The request id generation function.
+ *
+ * @returns {MiddlewareHandler} The middleware handler function.
+ *
+ * @example
+ * ```ts
+ * type Variables = RequestIDVariables
+ * const app = new Hono<{Variables: Variables}>()
+ *
+ * app.use(requestID())
+ * app.get('/', (c) => {
+ *   console.log(c.get('requestID')) // Debug
+ *   return c.text('Hello World!')
+ * })
+ * ```
+ */
+export const requestID = (options?: RequesIDOptions): MiddlewareHandler => {
+  const limitLength = options?.limitLength ?? 255
+  const headerName = options?.headerName ?? 'X-Request-Id'
+
+  return async function requestID(c, next) {
+    // If `headerName` is empty string, req.header will return the object
+    let requestId = headerName ? c.req.header(headerName) : undefined
+
+    if (requestId) {
+      requestId = requestId.replace(/[^\w\-]/g, '')
+      requestId = limitLength > 0 ? requestId.substring(0, limitLength) : requestId
+    } else {
+      requestId = options?.generator?.() ?? crypto.randomUUID()
```

Comment:
Note that, `crypto.randomUUID()` may not be a stable method to use if Hono plan to focus on [supporting Web Browsers](https://x.com/honojs/status/1714947374886637883)

Reply:
Ah, I also worried about `crypto.randomUUID()`. Either way, I think `crypto.randomUUID()` works on major browsers: https://developer.mozilla.org/en-US/docs/Web/API/Crypto/randomUUID

Or am I missing something?

Reply:
I think it's one of web-standards, so I don't think there's a problem.
One thing to note is that this does not work in Node v18. However, there are `generator` option for avoiding that.

Reply:
`crypto.randomUUID()` only works in HTTPS @yusukebe @ryuapp I'm afraid it something we should thinking of

Reply:
Also a random stackoverflow post I've found talking about it: https://stackoverflow.com/a/77981051/16258067

Reply:
Yup, `crypto.randomUUID()` doesn't work on this domain http://example.com/. But I believe the majority of the apps will be deployed on `HTTPS` so this shouldn't be a problem.

But if you want, We can do an `if else` statement to check if `randomUUID` is present in `crypto` and can either show a warning or do a simple implementation like `() => "add-a-generator"`, which doesn't require `crypto`

Reply:
> crypto.randomUUID() only works in HTTPS

I think this problem can be ignored.
It can be assumed that only Hono playgrounds connected via HTTP are affected.
That's a very rare case.
As mentioned above, if `crypto.randomUUID()` is not possible, we can use `generator` option.

All we have to do is write the cautionary point in the docs.

Reply:
@fzn0x @MathurAditya724 Ahh, thanks!

Does this mean that if I run an app with this Request ID middleware on localhost as a Service Worker, it does not work well? Hmm, we don't want to use anything that isn't common as much as possible.

Reply:
However, creating our own UUID generation logic in the `hono` package might be wasteful, so maybe the user can just pass it as an option.

Reply:
> Does this mean that if I run an app with this Request ID middleware on localhost as a Service Worker, it does not work well?

It works well on localhost. Only 'HTTP' connection is a problem. Check it there:
https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts#when_is_a_context_considered_secure

> Locally-delivered resources such as those with http://127.0.0.1 URLs, http://localhost and http://*.localhost URLs (e.g. http://dev.whatever.localhost/), and file:// URLs are also considered to have been delivered securely.

Although it is not necessary to use UUID, I did it this way because it is minimum codes and many frameworks use UUIDv4 for request ID middleware.

Reply:
> It works well on localhost. Only 'HTTP' connection is a problem. Check it there:

Thanks, I understood well. So I think it's using `crypto.randomUUID()` is good! Shall we go with this?

Reply:
@yusukebe Lets ship it and wait how users will react! 🚀